assets/js/googlesitekit/data/create-fetch-store.js

```diff
@@ -0,0 +1,270 @@
+
+	const FETCH = `FETCH_${ constantBaseName }`;
+	const START_FETCH = `START_${ FETCH }`;
+	const FINISH_FETCH = `FINISH_${ FETCH }`;
+	const CATCH_FETCH = `CATCH_${ FETCH }`;
+	const RECEIVE = `RECEIVE_${ constantBaseName }`;
+
+	const fetchCreator = `fetch${ pascalCaseBaseName }`;
+	const receiveCreator = `receive${ pascalCaseBaseName }`;
+	const isFetching = `isFetching${ pascalCaseBaseName }`;
+
+	const INITIAL_STATE = {
+		[ isFetching ]: {},
+	};
+
+	const actions = {
+		*[ fetchCreator ]( ...args ) {
+			let response;
+			let error;
+			let params;
+
+			try {
+				params = argsToParams( ...args );
+			} catch ( err ) {
+				// Parameters should never be invalid here, this needs to be
+				// strict and inform the developer of the issue.
+				global.console.error( err.message );
+				error = err;
+				return { response, error };
+			}
+
+			yield {
+				payload: { params },
+				type: START_FETCH,
+			};
+
+			try {
+				response = yield {
+					payload: { params },
+					type: FETCH,
+				};
+
+				yield actions[ receiveCreator ]( response, params );
+
+				yield {
+					payload: { params },
+					type: FINISH_FETCH,
+				};
+			} catch ( e ) {
+				error = e;
+
+				yield {
+					payload: { error, params },
+					type: CATCH_FETCH,
+				};
+			}
+
+			return { response, error };
+		},
+
+		[ receiveCreator ]( response, params ) {
+			invariant( response !== undefined, 'response is required.' );
+			if ( requiresParams ) {
+				invariant( isPlainObject( params ), 'params is required.' );
+			} else {
+				params = {};
+			}
+
+			return {
+				payload: { response, params },
+				type: RECEIVE,
+			};
+		},
+	};
+
+	const controls = {
+		[ FETCH ]: ( { payload } ) => {
+			return controlCallback( payload.params );
+		},
+	};
+
+	const reducer = ( state, { type, payload } ) => {
+		switch ( type ) {
+			case START_FETCH: {
+				const { params } = payload;
+				return {
+					...state,
+					[ isFetching ]: {
+						...state[ isFetching ],
+						[ stringifyObject( params ) ]: true,
+					},
+				};
+			}
+
+			case RECEIVE: {
+				const { response, params } = payload;
+				return reducerCallback( state, response, params );
+			}
+
+			case FINISH_FETCH: {
+				const { params } = payload;
+				return {
+					...state,
+					[ isFetching ]: {
+						...state[ isFetching ],
+						[ stringifyObject( params ) ]: false,
+					},
+				};
+			}
+
+			case CATCH_FETCH: {
+				const { error, params } = payload;
+				return {
+					...state,
+					error,
+					[ isFetching ]: {
+						...state[ isFetching ],
+						[ stringifyObject( params ) ]: false,
+					},
+				};
+			}
+
+			default: {
+				return { ...state };
+			}
+		}
+	};
+
+	const selectors = {
+		[ isFetching ]: ( state, ...args ) => {
+			if ( state[ isFetching ] === undefined ) {
+				return false;
+			}
+
+			let params;
+			try {
+				params = argsToParams( ...args );
+			} catch ( err ) {
+				// If parameters are invalid, fail silently here. It likely is
```

Comment:
Minor documentation tweak here: arguably it's not really "failing" silently here. This is expected behaviour because the arguments provided aren't yet valid because other selectors aren't ready.

That said, I'd expect `argsToParams` to handle invalid params accordingly and return an appropriately empty object or whatever if dependent selectors pass it `undefined` or `null` values.

Because this code catches _any_ error in `argsToParams` and will silence it and ignore it, it's potentially dangerous and obscuring real errors. It's a bit more work for developers, but it might be better to have `argsToParams` have to handle empty arguments than take this approach. Maybe we should ditch this try/catch block.

Reply:
That's a fair point, but I think we can go with this for now, let's see if we run into issues because of it. `argsToParams` should be fairly basic, just check for required parameters and validate as needed. Since this is only a (usually internal) selector to check whether a request is in progress, I'd even more so say the problem vector here is rather low.